project/OsgiCheck.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.io.File
+import sbt.MessageOnlyException
+import scala.collection.mutable.Map
+
+object OsgiCheck {
+  val prog = "osgiCheck"
+
+  // Directories in the repository root that are libraries which should be checked
+  val libDirMatchers = Seq("""^.*(daffodil-\S+)$""".r)
+
+  // Only filenames that match here are considered
+  val fileMatcher = """.*\.(scala|java)""".r
+
+  // Ignore these libraries, packages, and directory names
+  var ignoreLibs = Seq(
+    "daffodil-macro-lib", // multi-package
+  )
+  val ignorePackages = Set("org.apache.daffodil")
+  var ignoreDirs = Set("test")
+
+  // Statement that identifies package membership
+  val packageMatcher = """^\s*package (\S+);?.*$""".r
+
+  // Keeps a mapping of (K, V) where V is a library name and K is a package known to be
+  // owned by that library.
+  val owners: Map[String, String] = Map()
+  var nErrors = 0
+  var nWarnings = 0
+
+  def error(msg: String) = {
+    println(s"${prog}: error: ${msg}")
+    nErrors += 1
+  }
+
+  def warning(msg: String) = {
+    println(s"${prog}: warning: ${msg}")
+    nWarnings += 1
+  }
+
+  def checkPackage(fileName: String, libName: String, packageName: String): Unit = {
```

Comment:
The way to handle this logic in a per subproject task could be to make two tasks. One is something like "osgiOwnedPackages" and it returns a `Seq[String]` of packages that each subproject owns. Then an "osgiCheck" task would just access that task from each subproject and make sure there are no overlaps.